### ISSUE TITLE
ONNX is slower than Pytorch

### ISSUE THREAD

**Issue by YOOHYOJEONG** [ultralytics/ultralytics#17127]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, 
I did benchmark COCO through the command below using pre-trained model.

`yolo benchmark model=/ultralytics/DATA/weights/coco_pretrained/yolov8/yolov8n.pt data=/ultralytics/DATA/weights/coco_pretrained/coco.yaml   
`      

The result is as shown in the image below.   
<img width="508" alt="image" src="https://github.com/user-attachments/assets/1f3fee18-c67b-4e3f-b695-94a4da5a5865">

I think the result is a little strange here.
I know that onnx is usually faster than pytorch, but the benchmark result is the opposite.   
   
   
I'm using ultralytics through docker image(https://hub.docker.com/layers/ultralytics/ultralytics/latest/images/sha256-2d498f6677bcc7ccaf5ebda1c3091e1c683e10efe82591be47b3ec1904c86f02?context=explore).

`docker pull ultralytics/ultralytics:latest`   

So, my ultralytics version is "8.3.18"
   

May I know why onnx is slower than pytorch in your docker image environment?

`yolo checks`
Ultralytics 8.3.18 🚀 Python-3.11.9 torch-2.4.1+cu121 CUDA:0 (NVIDIA GeForce RTX 4090, 24217MiB)
Setup complete ✅ (32 CPUs, 125.5 GB RAM, 1549.8/1832.2 GB disk)

OS                  Linux-6.8.0-45-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.11.9
Install             git
RAM                 125.51 GB
Disk                1549.8/1832.2 GB
CPU                 Intel Core(TM) i9-14900K
CPU count           32
GPU                 NVIDIA GeForce RTX 4090, 24217MiB
GPU count           2
CUDA                12.1

numpy               ✅ 1.23.5>=1.23.0
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.2.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.4.1+cu121>=1.8.0
torchvision         ✅ 0.19.1+cu121>=0.9.0
tqdm                ✅ 4.66.5>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.9>=2.0.0
torch               ✅ 2.4.1+cu121!=2.4.0,>=1.8.0; sys_platform == "win32"

I installed onnx too.

![image](https://github.com/user-attachments/assets/f474bac6-f20b-4471-a7e9-a447223c66ae)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @YOOHYOJEONG, thank you for your interest in Ultralytics 🚀 and for bringing up this performance observation! This is an automated response. An Ultralytics engineer will assist you soon.

We recommend checking out the [Ultralytics Docs](https://docs.ultralytics.com) for more insights into using our tools efficiently. You can explore various [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples that might answer some of your queries.

If this is a 🐛 Bug Report or performance-related concern, it would be helpful to provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us understand the issue better.

To ensure that you're using the latest features and improvements, we suggest upgrading to the most recent version of the `ultralytics` package:

```bash
pip install -U ultralytics
```

In case you are eager to dive deeper or have discussions with the community, feel free to connect through [Discord](https://ultralytics.com/discord) 🎧 for real-time interaction, or engage in detailed discussions on our [Discourse](https://community.ultralytics.com). For broader community engagement, our [Subreddit](https://reddit.com/r/ultralytics) is a great place to go.

## Verified Environment Options

Here are some environments where you can run YOLO with everything set up:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, it indicates that all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing, ensuring smooth operations across multiple environments and tasks.

Stay connected for more updates and feel free to share any additional details that can help us assist you better!

**Comment by Y-T-G** [comment]
I don't think it uses GPU

**Comment by YOOHYOJEONG** [comment]
@Y-T-G 
I saw an increase in gpu usage when using onnx.
I actually checked the time difference between when using gpu and when not using it.

It uses gpu.

**Comment by Y-T-G** [comment]
The code only passes CPU

https://github.com/ultralytics/ultralytics/blob/8f0a94409fb2f6320b2d42db9feb4dea7ec40ac1/ultralytics/utils/benchmarks.py#L489

**Comment by YOOHYOJEONG** [comment]
Yes,
I checked it and modified it with CUDAExecutionProvider.
Still, onnx is slower than pytorch with the same result.

![image](https://github.com/user-attachments/assets/a5769819-11b2-44b3-93f9-242345898ea8)

**Comment by ambitious-octopus** [comment]
You are using a container, are you passing the `--gpus all` flag when you start the container?

**Comment by YOOHYOJEONG** [comment]
@ambitious-octopus 
Yes. I did..

**Comment by Y-T-G** [comment]
I don't think ONNX is faster than PyTorch. Even in bare bones inference, ONNX is slower.

```python
In [45]: %timeit onnx_model.predictor.model.session.run(onnx_model.predictor.model.output_names, {'images': prep_c})
8.18 ms ± 7.96 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
In [47]: %timeit _ = pt_model.predictor.inference(prep)
6.43 ms ± 16.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by ambitious-octopus** [comment]
Also onnx is using `batch=1` not sure about the vanilla Pytorch batch, I guess higher (like 16). cc @Laughing-q

**Comment by YOOHYOJEONG** [comment]
Okay...
Is there any data comparing the speed of pytorch and onnx using ultralytics benchmark?

And,
In general, I know that if I export the pytorch model to onnx, the model size will also be much smaller,

In the ultralytics benchmark results, the model size of onnx is much larger than pytorch.  
![image](https://github.com/user-attachments/assets/b2413220-3c7d-4147-be2d-77f00b21ca30)
   

I think this is weird, too. Is this right?

**Comment by YOOHYOJEONG** [comment]
@ambitious-octopus 

I confirmed that both onnx and pytorch use `batch=1` when the benchmark is conducted...

**Comment by YOOHYOJEONG** [comment]
@Y-T-G 

Do you think it's normal?
Usually, you have to be faster faster and light.

Is it a model that transformed into an extensions of multilingual?

**Comment by sunnyYaSky** [comment]
I also have this problem .I test .pt and  .onnx model  and find that  .pt is faster than .onnx  on gpu and cpu .this is my result :
<html>
<body>
<article data-content = '[{&quot;name&quot;:&quot;table&quot;,&quot;type&quot;:&quot;block&quot;,&quot;data&quot;:{&quot;colsWidth&quot;:[220.66666666666666,220.66666666666666,220.66666666666666],&quot;rowsHeight&quot;:[40,40,40]},&quot;nodes&quot;:[{&quot;name&quot;:&quot;table-row&quot;,&quot;type&quot;:&quot;block&quot;,&quot;nodes&quot;:[{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;table-cell&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;paragraph&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;text&quot;,&quot;leaves&quot;:[{&quot;text&quot;:&quot;n&nbsp;模型&quot;,&quot;marks&quot;:[]}]}]}]},{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;table-cell&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;paragraph&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;text&quot;,&quot;leaves&quot;:[{&quot;text&quot;:&quot;cpu(ms)&quot;,&quot;marks&quot;:[]}]}]}]},{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;table-cell&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;paragraph&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;text&quot;,&quot;leaves&quot;:[{&quot;text&quot;:&quot;gpu（ms）&quot;,&quot;marks&quot;:[]}]}]}]}]},{&quot;name&quot;:&quot;table-row&quot;,&quot;type&quot;:&quot;block&quot;,&quot;nodes&quot;:[{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;table-cell&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;paragraph&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;text&quot;,&quot;leaves&quot;:[{&quot;text&quot;:&quot;yolov11.pt&quot;,&quot;marks&quot;:[]}]}]}]},{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;table-cell&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;paragraph&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;text&quot;,&quot;leaves&quot;:[{&quot;text&quot;:&quot;43&quot;,&quot;marks&quot;:[]}]}]}]},{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;table-cell&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;paragraph&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;text&quot;,&quot;leaves&quot;:[{&quot;text&quot;:&quot;10&quot;,&quot;marks&quot;:[]}]}]}]}]},{&quot;name&quot;:&quot;table-row&quot;,&quot;type&quot;:&quot;block&quot;,&quot;nodes&quot;:[{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;table-cell&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;paragraph&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;text&quot;,&quot;leaves&quot;:[{&quot;text&quot;:&quot;yolov11.onnx&quot;,&quot;marks&quot;:[]}]}]}]},{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;table-cell&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;paragraph&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;text&quot;,&quot;leaves&quot;:[{&quot;text&quot;:&quot;45&quot;,&quot;marks&quot;:[]}]}]}]},{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;table-cell&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;block&quot;,&quot;name&quot;:&quot;paragraph&quot;,&quot;data&quot;:{},&quot;nodes&quot;:[{&quot;type&quot;:&quot;text&quot;,&quot;leaves&quot;:[{&quot;text&quot;:&quot;42&quot;,&quot;marks&quot;:[]}]}]}]}]}]}]' ><div yne-bulb-block="table" style="overflow: auto;">
            
n 模型 | cpu(ms) | gpu（ms）
-- | -- | --
yolov11.pt | 43 | 10
yolov11.onnx | 45 | 42


        </div></article>
</body>
</html>

**Comment by Y-T-G** [comment]
I noticed inference with `opset=17` ONNX is much faster on CUDA compared to other opsets

![Image](https://github.com/user-attachments/assets/e208b9c9-043f-4adb-ad01-99247e7696d5)
